File: ruby/017_CrossingSegments.rb
# 017 - Crossing Segments（★7）
# 余事象
# BIT(FenwickTree)

class FenwickTree
  attr_accessor :data, :size

  def initialize(arg)
    case arg
    when Array
      @size = arg.size
      @data = [0].concat(arg)
      1.upto(@size) do |i|
        up = i + (i & -i)
        next if up > @size
        @data[up] += @data[i]
      end
    when Integer
      @size = arg
      @data = Array.new(@size + 1, 0)
    end
  end

  def add(pos, value)
    idx = pos + 1
    while idx <= size
      data[idx] += value
      idx += idx & -idx
    end
  end

  # [l, r)
  def sum(l, r=nil)
    r ? _sum(r) - _sum(l) : _sum(l)
  end

  def _sum(r)
    idx = r
    res = 0
    while idx > 0
      res += data[idx]
      idx -= idx & -idx
    end
    res
  end
end

N, M = gets.split.map(&:to_i)
LR = []
COUNT = Array.new(N, 0)
R = Array.new(N, 0)
M.times do
  l, r = gets.split.map(&:to_i).map(&:pred)
  LR << [l, r]
  COUNT[l] += 1
  COUNT[r] += 1
  R[r] += 1
end

# 全体の場合の数
total = M * (M - 1) / 2

# ① 端点で交わる場合の数
comp1 = COUNT.sum { _1 * (_1 - 1) / 2 }

# ② Rs < Ltの場合の数
rcsum = R.reduce([0]) {|acc, i| acc << acc[-1] + i }
comp2 = LR.sum {|l, _| rcsum[l] }

# ③ 片方の区間内に入る場合の数
L = Array.new(N, 0)
lcsum = FenwickTree.new(N)
comp3 = 0
LR.sort_by { [_1[1], _1[0]] }.each do |l, r|
  comp3 += lcsum.sum(l + 1, r)
  lcsum.add(l, 1)
end

puts total - comp1 - comp2 - comp3

File: kotlin/017_CrossingSegments.kt
/**
 * 017 - Crossing Segments（★7）
 * 余事象
 * BIT(FenwickTree)
 */

class FenwickTree private constructor(
  private val data: LongArray,
  private val size: Int
)  {
  companion object {
    fun of(size: Int) = FenwickTree(LongArray(size + 1) {0}, size)
    fun of(list: List<Long>) = FenwickTree(longArrayOf(0) + list.toLongArray(), list.size)
  }

  fun add(pos: Int, value: Long) {
    var idx = pos + 1
    while (idx <= size) {
      data[idx] = data[idx] + value
      idx += idx.and(-idx)
    }
  }

  fun sum(l: Int, r: Int): Long {
    return _sum(r) - _sum(l)
  }

  fun _sum(r: Int): Long {
    var idx = r
    var res = 0L
    while (idx > 0) {
      res += data[idx]
      idx -= idx.and(-idx)
    }
    return res
  }
}


fun main() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val LR = mutableListOf<Pair<Int, Int>>()
  val COUNT = LongArray(N) { 0 }
  val R = LongArray(N) { 0 }
  repeat(M) {
    val (l, r) = readLine()!!.split(" ").map(String::toInt).map(Int::dec)
    LR.add(Pair(l, r))
    COUNT[l]++
    COUNT[r]++
    R[r]++
  }

  val total = M.toLong() * (M - 1) / 2

  // ① 点を共有
  val comp1 = COUNT.sumOf { it * (it - 1) / 2 }

  // ② Rs < Lt
  val rcsum = R.scan(0, Long::plus)
  val comp2 = LR.sumOf { (l, _) -> rcsum[l] }

  // ③ 内包
  val L = FenwickTree.of(N)
  var comp3 = 0L
  LR.sortedWith(compareBy({ it.second }, { it.first })).forEach { (l, r) ->
    comp3 += L.sum(l + 1, r)
    L.add(l, 1)
  }

  println(total - comp1 - comp2 - comp3)
}
